### PR TITLE
Migrate iterators.py for datatree.

### PR DESCRIPTION
**Pull request by owenlittlejohns**
This PR continues the overall work of [migrating DataTree into xarray](https://github.com/pydata/xarray/issues/8572).

`iterators.py` does not have direct tests. In discussions with @TomNicholas and @flamingbear, we concurred that other unit tests utilise this functionality.

- [x] Closes migration step for `iterators.py` #8572
- [ ] ~~Tests added~~
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

File: xarray/core/iterators.py
Comment (owenlittlejohns) [inline xarray/core/iterators.py:5]: This change looked a bit unexpected, but `typing.Iterator` is [a deprecated alias for collections.abc.Iterator](https://docs.python.org/3/library/typing.html#typing.Iterator)

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:22]: Odd format of this docstring, xarray usually follows the numpy docstring format.

Adding some examples how it's used would be nice as well. Then you would also get some doctests for free.

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:14]: Considering this `__init__`, using `@dataclass` could be an option here.

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:66]: Add typing to the rest of the arguments? They should probably have the same typing as above right?

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:73]:
```suggestion
    def _get_children(children: list[Tree], stop: Callable) -> list[Tree]:
```
?

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py]: I've migrated this documentation string up to the class level, which seems in keeping with the rest of the repository. I've then added examples for the subclasses of this abstract base class to those subclasses themselves.

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py]: Good call. Done in: [dc8ad9](https://github.com/pydata/xarray/pull/8879/commits/dc8ad955173effde954f47c3b380fc5bc4a7432c)

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py]: Also done in: [dc8ad9](https://github.com/pydata/xarray/pull/8879/commits/dc8ad955173effde954f47c3b380fc5bc4a7432c)

**Comment by TomNicholas** [comment]
As this module comes directly from the [anytree library](https://github.com/c0fec0de/anytree/tree/main/anytree/iterators), we might need to copy its license into [xarray/licenses](https://github.com/pydata/xarray/tree/main/licenses).

File: xarray/core/treenode.py
Comment (TomNicholas) [inline xarray/core/treenode.py:338]: So we currently have two patterns of iteration used: depth-first (`PreOrderIter`) and breadth-first (`LevelOrderIter`). It looks like `PreOrderIter` is used in `TreeNode.subtree`, and `LevelOrderIter` is used in `mapping.diff_treestructure`. `diff_treestructure` is called inside `check_isomorphic`, and every other time we iterate over the tree it just calls `.subtree`.

Why the distinction? In `diff_treestructure` when comparing two trees with multiple points at which their structure deviates, the order of iteration will affect which deviation is raised as an error first. @flamingbear and I think here is it more intuitive to raise errors from the top-level of the tree first, so we agree that `LevelOrderIter` is the right choice here.

For mapping over the nodes of the tree in `.subtree`, I had previously thought that there was a good reason to use `PreOrderIter`. My reasoning was to do with how `dt['/a/b/c/d/'] = data` would immediately create the nodes `/a`, `/a/b`, `/a/b/c` even if they didn't already exist. But actually I don't think that's relevant here.

Another reason to use `PreOrderIter` might be around using `map_over_subtree` to map an operation that then fails. For example taking `dt.mean(dim='time')` when some of the nodes doesn't have a time dimension. Again the order of iteration affects which node will raise the first error.

However, if we actually think that `LevelOrderIter` is fine for the `.subtree` case too, we can potentially simplify this PR considerably by replacing the whole `AbstractIter/PreOrderIter/LevelOrderIter` framework with just a single `iterate_breadth_first` function that returns an iterator, and use that in all cases.

Comment (TomNicholas) [reply xarray/core/treenode.py:338]: (Also then we wouldn't need the anytree license)

Comment (owenlittlejohns) [reply xarray/core/treenode.py]: @TomNicholas - I've made some updates (or rather nabbed some work @flamingbear did and made some `mypy` tweaks).

A couple of things here:

* This now contains the single class, but hasn't wrapped it in a function. I think it gets at what you were after, but just FYI. (Although, part of me wonders if this now warrants a module all of it's own?)
* I left the `anytree` license in the PR, because the `LevelOrderIter` is still 99% the code from `anytree`.

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py:65]: @Illviljan - sorry for not replying to this comment before. In the latest PR, the `AbstractIter` class was removed in favour or importing directly from `anytree`, so I think this comment is now addressed. Let me know if not, though.

Comment (Illviljan) [reply xarray/core/iterators.py:65]: If you want, `LevelOrderIter` can still use `@dataclass`.

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:14]:
```suggestion
    """
    Iterate over tree applying level-order strategy starting at `node`.
    This is the iterator used by `DataTree` to traverse nodes.
```

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:28]:
```suggestion

    Examples
```

File: xarray/core/iterators.py
Comment (Illviljan) [inline xarray/core/iterators.py:63]:
```suggestion
    """
```

File: xarray/tests/test_treenode.py
Comment (owenlittlejohns) [reply xarray/tests/test_treenode.py:323]: With the values of `expected`, I'm guessing that because it's all just built-in Python types (a `list` or `str` values) `mypy` can understand that ([here is a random similar example from test_dataset.py](https://github.com/pydata/xarray/blob/main/xarray/tests/test_dataset.py#L3481).

For the places where `create_test_tree` is called (so `root` in this test), the typing is done within that test function, so my guess is that `mypy` is pulling the typing from there (see L242 - L250).

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py]: I made the suggested whitespace changes in [this commit](https://github.com/pydata/xarray/pull/8879/commits/4ca733947cd72f2a01dd7f20b00d314600a97918).

Well I made 2.5 of them - I left the first line of the class documentation string for `LevelOrderIter` on the same line, as this seems consistent with other documentation strings in the repository. I did de-dent the second line, though, because that definitely looked a bit off.

File: xarray/core/iterators.py
Comment (owenlittlejohns) [reply xarray/core/iterators.py:65]: I wasn't familiar with this decorator prior to this PR, but it looks like it generates some magic methods for the class. I'm a little wary of adding them (for example, do we want `LevelOrderIter.__eq__`?). Is there a strong argument here for this class needing them?

File: xarray/tests/test_datatree.py
Comment (TomNicholas) [inline xarray/tests/test_datatree.py:498]: Nice. Interesting that this is the only test that explicitly checks the order of iteration.

File: xarray/tests/test_treenode.py
Comment (TomNicholas) [reply xarray/tests/test_treenode.py:323]: Let's leave worrying about mypy on tests for #8926

**Review by TomNicholas — APPROVED** [review]
This looks good to me.

File: xarray/core/treenode.py
Comment (TomNicholas) [reply xarray/core/treenode.py]: > This now contains the single class, but hasn't wrapped it in a function.

That's fine, thanks.

> (Although, part of me wonders if this now warrants a module all of it's own?)

If you wanted to move it to `treenode.py` instead that would also make sense. But I don't have a strong opinion, and it's trivial to change later.

Comment (owenlittlejohns) [reply xarray/core/treenode.py]: Okay - I'm happy to leave this in it's own module for now.

**Comment by flamingbear** [comment]
This looks good to me now too @TomNicholas are you waiting for another approval?

**Pull request merged**
